Adds **kwargs parameter to password_hash to support more algorithms.

mfuchs opened the pull request:
##### SUMMARY
Adds **kwargs parameter to password_hash to support more algorithms.
    
This adds support for the bcrypt algorithm as described in the documentation.
The msdcc and msdcc2 algorithm are also supported now.

Fixes #45392

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
utils/encrypt

##### ANSIBLE VERSION
```paste below
ansible 2.7.0rc2.post0
```

mfuchs committed:
Adds support to specify the algorithm to use for bcrypt.

The ident parameter, as described in the documentation, was not supported.

mscherer commented:
Please do separate PR if you start to do multiple change, or at least, multiple commits.

mscherer commented on lib/ansible/utils/encrypt.py:
```diff
@@ -104,7 +104,10 @@ def _rounds(self, rounds):
-    def _hash(self, secret, salt, rounds):
+    def _hash(self, secret, salt, rounds, **kwargs):
+        if kwargs:
+            raise AnsibleError("crypt.crypt does not support paramters '%s'" % str(kwargs))
```
s/paramters/parameters/

mscherer commented on lib/ansible/utils/encrypt.py:
```diff
@@ -170,7 +174,10 @@ def _hash(self, secret, salt, salt_size, rounds):
         if hasattr(self.crypt_algo, 'hash'):
-            result = self.crypt_algo.using(**settings).hash(secret)
+            if self.algorithm in ('msdcc', 'msdcc2'):
+                result = self.crypt_algo.hash(secret, **settings)
```
So, this doesn't verify anymore that user is set, unlike the previous PR, I think that should be corrected.

mfuchs replied:
It is done implicitly by passlib.
So if the user is not set an exception would be thrown, similar to what https://github.com/ansible/ansible/issues/45392 experienced.

This is the same behaviour as for all other hashing algorithms.
I.e. the checking of what parameters are supported/needed is done by passlib.

mscherer replied:
Throwing a exception do not seems like great UX for users.

mfuchs replied:
I think from a maintenance point it is easier to let passlib handle the parameters.
Otherwise we would have to take care of each algorithm ourselves and have a list of required and optional parameters for them.
It becomes worse if algorithms retrieve new paramters that we already support.
Now iff new algorithms are added to passlib we support them automatically.
Given that they use the 'using' and 'hash' functions.

abadger commented:
I would add the parameters.  Kwargs don't actually mean we haven't added
the parameters... It just means that we haven't made those parameters easy
to see and easy to error check.  If it doesn't make sense to add more
parameters explicitly then it would also not make sense to add them via
kwargs.

On Sun, Sep 16, 2018, 7:19 AM Matthias Fuchs <notifications@github.com>
wrote:

> @abadger <https://github.com/abadger> I know you were not a fan of
> **kwargs in #21215 <https://github.com/ansible/ansible/pull/21215>.
> But I am not sure if it is good to add even further parameters to the
> functions (i.e. 'user' and 'ident').
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ansible/ansible/pull/45692#issuecomment-421772813>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAMxWti0TMZwabwws2Btbq9bRpqP9AUiks5ubl36gaJpZM4Wq3wB>
> .
>

mfuchs commented:
@mscherer do you mean splitting it up into two commits: One adding bcrypt support and the other for msdcc/msdcc/2?

@abadger ok, I removed **kwargs again. :)

mscherer commented:
Yes, so it is easier to see the meaning of each, and review them.

mfuchs committed:
Adds support for the msdcc/msdcc2 algorithms.

This is done by adding the 'user' parameters for msdcc/msdcc2.
Furthermore msdcc/msdcc2 does not support the 'using' function with passlib
1.7 and newer.
Thus there is specific handling if these algorithms are used.

Fixes #45392

mfuchs commented:
@mscherer I split the commit into two parts.

samccann commented:
@mfuchs Sorry for the delays. Can you rebase and resolve any open review comments on this PR?

Thanks!

mfuchs commented:
@samccann rebased and also resolved all comments.

samccann commented:
@mscherer @abadger - can you take another look at this PR to see if everything has been addressed? (and merge if you approve?)  Thanks!

mscherer commented on docs/docsite/rst/user_guide/playbooks_filters.rst:
```diff
@@ -848,6 +848,13 @@ Some hash types allow providing a rounds parameter::
     {{ 'secretpassword' | password_hash('sha256', 'mysecretsalt', rounds=10000) }}
 
+When`Passlib <https://passlib.readthedocs.io/en/stable/>`_ is installed
+`password_hash` supports more crypt schematas::
```
Small typo (not blocking, but since I review)

mscherer commented:
You should also add a changelog fragment (IIRC the new merging rule, I am a bit out of touch)

tumluliu commented:
Hi @mfuchs , thanks a lot for the nice job! Is there any news that when this can be merged? The `ident` param is very required in our use cases

bcoca commented:
`ident` was already added in parallel, if you still wish to add 'user' parameter please update the PR (solve conflicts) and add a changelog.

waiting_on_contributor
